code/game/machinery/vending.dm

```diff
@@ -120,31 +114,103 @@
 	QDEL_NULL(inserted_item)
 	return ..()
 
+/obj/machinery/vending/RefreshParts()         //Better would be to make constructable child
+	if(!component_parts)
+		return
+
+	product_records = list()
+	hidden_records = list()
+	coin_records = list()
+	build_inventory(products, product_records, start_empty = TRUE)
+	build_inventory(contraband, hidden_records, start_empty = TRUE)
+	build_inventory(premium, coin_records, start_empty = TRUE)
+	for(var/obj/item/vending_refill/VR in component_parts)
+		restock(VR)
+
+
 /**
  *  Build src.produdct_records from the products lists
  *
  *  src.products, src.contraband, src.premium, and src.prices allow specifying
  *  products that the vending machine is to carry without manually populating
  *  src.product_records.
  */
-/obj/machinery/vending/proc/build_inventory()
-	var/list/all_products = list(
-		list(products, CAT_NORMAL),
-		list(contraband, CAT_HIDDEN),
-		list(premium, CAT_COIN))
-
-	for(var/current_list in all_products)
-		var/category = current_list[2]
+/obj/machinery/vending/proc/build_inventory(list/productlist, list/recordlist, start_empty = FALSE)
+	for(var/typepath in productlist)
+		var/amount = productlist[typepath]
+		if(isnull(amount))
+			amount = 0
+
+		var/atom/temp = typepath
+		var/datum/data/vending_product/R = new /datum/data/vending_product()
+		R.name = initial(temp.name)
+		R.product_path = typepath
+		if(!start_empty)
+			R.amount = amount
+		R.max_amount = amount
+		R.price = (typepath in prices) ? prices[typepath] : 0

```

Comment:
I didn't feel like doing it in this PR, but we should probably do what TG does, and specify the price on the object, itself, then fetch it with `R.price = initial(temp.custom_price)`, then have a var set, on machines, that if `FALSE` makes everything on it free.